### ISSUE TITLE
Alternative use of RE.halt() causes data to be marked incorrectly

### ISSUE THREAD

**Issue by ambarb** [bluesky/bluesky#1296]
`RE.hault()`  is a great option to have, but the use case is more narrow than what is used in practice. If I get to the end of the measurement, I will use `^C` to keep the motors where they are instead of waiting the 5 minutes for everything to go back to the initial conditions.  I use halt to save time and  not to just prevent a motor crash.  In both of these cases, it doesn't mean that the data was bad, it just means I needed (or preferred to stop but not return the motors to any position).  

Is there any flexibility in re-thinking how the exit_status is labeled by `RE.halt()`.  Or could we add an argument to `RE.stop()` to keep the motors in place?  Perhaps  this second option is nice because you could also add that argument to `RE.abort()` and just get rid of `RE.halt()`

**Comment by tacaswell** [comment]
This is likely identifying a need, but I don't think the proposed solutions will work.

The core problem is that `halt` does not simply not return the motors, in prevents _any_ message from being further processed by the RunEngine and the plans have no way to influence what happens between `RE.halt()` and when you get the prompt back.  It really is a bit of breaking the glass and pulling the emergency brake so I don't think we should normalize it.

On the other hand `RE.stop()` throws a known exception into the plan which the plan can catch and fully suppress (as you may recall happened with one of the plans at SIX that caught a bit too aggressively and couldn't be stopped!).  This is what lets us do arbitrary clean up, like moving all of the motors back to where we found them.  However, because all of this is implemented via the plans, not via the RE, there is not a single place where we could change this.

Are the motors being moved back by one of our wrappers or via a custom plan?  If it is the later, and you always want to skip moving the motors back on `stop` you could do that via


```python

do_reset = True
try:
    yield from main_plan
except RequestStop:
    do_reset = False
    raise
except GeneratorExit:
     # because python says so
    do_reset = False 
    raise
finally:
    if do_reset:
        yield from reset_plan()
```

If you only sometimes want to not move the motors back, it would be possible to add arguements to `RE.stop` that could be injected into the RequestStop object that gets thrown into the plan which we could then do some logic on in the code to decide if we want to do the reset.

Another option would be to use a custom reset plan that looks at a PV you could set side-band to decide if you want to reset the motors or not.

In either case I want to understand a bit better why you need this and if there is a less-drastic solution we can use.

**Comment by danielballan** [comment]
> If you only sometimes want to not move the motors back, it would be possible to add arguements to RE.stop that could be injected into the RequestStop object that gets thrown into the plan which we could then do some logic on in the code to decide if we want to do the reset.

That sounds very useful—a way for the user to tell the plan interactively _how_ to clean up using whatever arbitrary parameter make sense to that plan—passed through RE.stop(...) and from there into the plan via the exception.

**Comment by ambarb** [comment]
> Are the motors being moved back by one of our wrappers or via a custom plan? 

I am referring only to bluesky plans here for this "inspiration" to the enhancement request.  Say you did a scan in reciprocal space that rotated phi more than 180 degrees and moved chi my more than 70 degrees.  You would not want to go all the way back to the beginning, but it would be bad to mark the scan as aborted (otherwise, what is the point to mark as aborted if good data versus probably meaningless data can both be aborted).  Of course, if one only uses absolute scans, then maybe this isn't a problem, but I don't think we should be forcing people to use absolute scans if they want the ability to stop a scan without moving motors back to the beginning of the scan.

For the custom scans, I agree, people should add the clean up stuff themselves.  It might be nice to have a reserved cleanup plan for not moving the motors so you would just need to import that bit into the custom scan plan.

**Comment by prjemian** [comment]
This issue looks similar to https://github.com/APS-USAXS/ipython-usaxs/issues/255, which resets the instrument when a scan is ended abnormally.

> When we decide to stop and abort a plan, the instrument is not reset to starting condition.

In that issue, @danielaballan said *to override the device's `stop(success)` method.*  We have not implemented that yet.

**Comment by tacaswell** [comment]
Currently the `rel_scan` plan uses a "make relative" and a "reset the position" decorators, is the simplest fix here to add a "move the motors back" flag to the plans?

Another reason that we don't want to use `RE.halt` regularly is that among the clean up messages that need to be run could be removing any local subscriptions or suspenders.  If you use `halt` those may not be pulled off leaving zombie callbacks/suspenders in the process and the RunEngine in an undefined state.   I think a good case should be made that if you use `halt`, the RE should be put in a panicked state and require a process restart before continuing.

In the case of detectors, customizing the `stop` method may work, however while technically possible, making the `stop` method on a motor move it, possibly a long distance, seems too counter-intuitive to be a good idea.  Further, the plan is what knows the context of where it started and where it should go back to (or not), I don't think we want the ophyd objects to have that much knowledge of the history of the motion etc.

I think a key question here is: Can we make the "go back or not" decision _before_ we start the scan or not?  

If we can make it before, then we can solve this using slightly smarter / more parameterized plans.  We can probably do this with no change to bluesky/ophyd (or minor backward-compatible changes to the "stock" plans).  

If the can only make this decision after we have hit Ctrl-C, then we will need to invert an additional communication protocol and conventions for controlling how cleanup happens based on the arguments to `stop` (which much more powerful, but also a bunch more work and will require changes to the core RunEngine).

**Comment by tacaswell** [comment]
>  It might be nice to have a reserved cleanup plan for not moving the motors so you would just need to import that bit into the custom scan plan.

The motors only move back if you explicitly opt-in to the behavior (which the `rel_*` scans do).

**Comment by ambarb** [comment]
Thanks for explaining how the `rel_*` works.  I did not realize that.  Does this mean that we can create modify the exception to include an if statement to prevent the motors from moving, but otherwise clean up the scan as normal when using `RE.stop()`?

>I think a key question here is: Can we make the "go back or not" decision before we start the scan or not?

This is the difference between measurements and experiments. So for some beam lines, the answer is yes, and for others the answer is no. 

> making the stop method on a motor move it, possibly a long distance, seems too counter-intuitive to be a good idea

I don't know why this seems counter-intuitive.  I do this all the time in sorting out what the alignment of a single crystal and what kind of reciprocal space scans that I want to launch (to sort out count rates and resolution).  Also,  if we actually get to the point of "live-streaming"  ptchography or tomography data (or at least being able to process the data after only some of it is collected), then people will want to stop the scan more often and potentially not move the motors back. Another example would be fixedQ Energy scans.  Now these are custom scans at this point, but who knows what will come out of pyhkl.  



>Another reason that we don't want to use RE.halt regularly 

I can say that if we don't have this functionality, then I will considered what you say about using `RE.halt()` too often, but honestly, I will have to decide if I want to keep track of scans with `exit_status` of `aborted` instead of `success` that I want to analyze.  The only hang up here is that I cannot use the full power of data broker (unless of course there is a way to append updates to the database).

Finally, I have to also say that I see users using `RE.halt()` in the way I have described.  As bluesky gets older and more people are using bluesky, they will certainly figure out.  However, for the users I see doing this now, they don't use the metadata to its full capability.   SO I don't think they realize their data gets marked as `aborted`.

**Comment by ambarb** [comment]
I should maybe explain more my strong preference.  There are some motors that require a backlash regardless of encoder because of mechanical reasons.  If the scan is allowed to stop at the end of a point, then the motor should be in an appropriate position.  IF the scan clean up returns motors to starting positions and user selects "STOP" on the CSS or PyDM or whatever equivalent screen, then the motor's read back position could be incorrect because the required backlash was not performed.  

I guess the larger questions is should I be using `aborted` value for the `exit_status` key to "pre-mark" data.  If that was not the intent, then this would be good to know.

**Comment by prjemian** [comment]
@ambarb wrote:

>  IF the scan clean up returns motors to starting positions and user selects "STOP" on the CSS or PyDM or whatever equivalent screen, then the motor's read back position could be incorrect because the required backlash was not performed.

Checked the documentation of the [EPICS motor record's `STOP` field](https://epics.anl.gov/bcda/synApps/motor/motorRecord.html#Fields_command), and found statement that is correct, even though I expected the motor record to make a *controlled stop* (accounting for backlash if necessary).  Confirmed this with MEDM control of a real motor using the EPICS motor record.

@kmpeters: I expect the motor record to account for backlash correction as part of a controlled stop.  This is worth a discussion on the motor records [GH repository](https://github.com/epics-modules/motor).  See: https://github.com/epics-modules/motor/issues/153

**Comment by kmpeters** [comment]
> @kmpeters: I expect the motor record to account for backlash correction as part of a controlled stop. This is worth a discussion on the motor records GH repository. See: epics-modules/motor#153

I've commented on the motor issue.  

I don't understand the case for a graceful stop.  Is it ever the case that the position a motor is in when it stops (controlled or not) is the exact place where it needs to be?  Wouldn't it always need to be reset to a prior position by a cleanup routine?  If the motor is exactly where it needs to be when it was stopped, why was it moving?

**Comment by prjemian** [comment]
@kmpeters - The use case is described in this [particular comment above](https://github.com/bluesky/bluesky/issues/1296#issuecomment-590449228).

>  There are some motors that require a backlash regardless of encoder because of mechanical reasons. If the scan is allowed to stop at the end of a point, then the motor should be in an appropriate position. IF the scan clean up returns motors to starting positions and user selects "STOP" on the CSS or PyDM or whatever equivalent screen, then the motor's read back position could be incorrect because the required backlash was not performed.

Since a possible backlash correction is not made after a STOP, the *reset* function is left to client software, in this case, the bluesky framework.  That opens the door to very inconsistent handling of what is easily a situation that is not rare.  

For a given motor, what special cases are there that should prevent a *controlled stop* from applying appropriate backlash correction?  The entire point of backlash correction is to leave the motion axis at a more predictable position after the move than if no backlash correction has been made.

**Comment by ambarb** [comment]
If we `^C` a scan with bluesky and it finishes collecting point, it has not yet moved the motors again so it is in a "valid, known" position.  To clean up the scan, we can `RE.stop()`, and the motors **_could_** return to the position prior to the scan.  Stopping this motion will create problems for many of our motors.

One motor we have, because of mechanical reasons, takes a very long time to travel, but if it starts to travel in the direction that needs the backlash, stopping it in CSS will put the sample outside of the beam (assuming small sample).  The motor itself is good within 1 micron as long as backlash is used in the backlash direction.

**Comment by kmpeters** [comment]
> For a given motor, what special cases are there that should prevent a controlled stop from applying appropriate backlash correction? The entire point of backlash correction is to leave the motion axis at a more predictable position after the move than if no backlash correction has been made.

How is the motor record supposed to know that the STOP=1 from ophyd means "do a controlled stop" while another STOP=1 from an APS beamline means "immediately stop the motor so it doesn't collide with something else"?

The response to the stop command is controller-specific.  Some controllers have allow the behavior to be customized.  Some will decelerate at a user-configurable deceleration rate.  Others will decelerate using the magnitude of the acceleration value.  Some consider it the same as pressing an emergency stop button and decelerate very quickly.

**Comment by tacaswell** [comment]
> I don't know why this seems counter-intuitive

There are two levels of `stop` here.  One is `RE.stop()` (which semi-forcefully exits the current plan), the other is `motor_obj.stop()` which for the most part just pushes the `STOP` field.  It makes perfect sense for in the clean up part of a scan it to move motors as far as it wants, etc.  I would find in very confusing that doing `motor_obj.stop()` caused it to move a long distance rather than stop where it is.   We have actually had issues with this on CSX with the piezos where on exiting the RE would call stop on everything we touched, unconditionally, to make sure that we don't accidentally leave any zombie motions.  This resulted in the piezos jumping because the `STOP` field on their controler cut the power and the stick-slip mechanism relaxed to the nearest stable point.

I guess an exception to this would be something like a hard-drive read head stowing itself when it is stopped to make it safe, but in cases like that I would expect all of the logic to be implemented below EPICS

I was specifically responding to 

> In that issue, @danielballan said to override the device's stop(success) method. We have not implemented that yet.

-----

> This is the difference between measurements and experiments. So for some beam lines, the answer is yes, and for others the answer is no.

In enough of the cases you need this functionality is it worth implementing the `no_reset` flag on all of the `rel_*` plans first?  Because it is only in the plans, that is something I am confident we can do quickly (and in a beamline profile if push comes to shove).  It is not developing anything new concepts or APIs, but using the tools we already have in a slightly different way.  Developing the feedback path from `RE.stop(....)` to the plans will require an order of magnitude more work because we would be adding a pretty significant new feature to the core of Bluesky 


----

> Finally, I have to also say that I see users using RE.halt() in the way I have described. As bluesky gets older and more people are using bluesky, they will certainly figure out. 

The more I think about it, the more I am concerned that this is leaving the RE is an undefined state.  Maybe we should panic the RE on `halt` and force a user who has pulled the emergency break to restart bsui.  If this has been done regularly, this could explain the reports that things seem to get slower over time (as you can be collecting an unbounded number of zombie callbacks) and could explain some of the seemingly impossible situations that SIX has hit.

**Comment by ambarb** [comment]
>The more I think about it, the more I am concerned that this is leaving the RE is an undefined state. 

I think you should force bsui restart if you don't want users using `RE.halt()` in the way I have described.  I spoke to Johnny and he indicates that SIX always uses `RE.abort()`, though users may do otherwise, which is where I saw users first using `RE.halt()`.  I myself only realized why last week when I was tempted to use it to fix the problem of waiting for motors to return - which is why we are having the conversation now. I don't usually explain `RE.halt()` for SIX users because the SIX scan plans are pretty set in stone (energy scan and just counting at specified energies and maybe sample positions).  There is no need for them to use it in my experience.  It would just be for staff during commissioning at this point to prevent a crash.

> In enough of the cases you need this functionality is it worth implementing the no_reset flag on all of the rel_* plans first? 

Are you saying to remove the motors returning to the starting position at the end of the plan?


>There are two levels of stop here. One is RE.stop() (which semi-forcefully exits the current plan), the other is motor_obj.stop() which for the most part just pushes the STOP field. 

Can you please clarify when `motor_obj.stop()` happens?  Is it that you don't want to use this to stop the motors from returning to the starting position by changing how `RE.stop()` works?

**Comment by tacaswell** [comment]
> Are you saying to remove the motors returning to the starting position at the end of the plan?

Yes, or at least making it optional.

> Can you please clarify when motor_obj.stop() happens?

We call `obj.stop` on every device we have called set on on the way out of the RE (unconditionally) and immediately when a run in interrupted (either via ctrl-C or via a suspender).   There is also a `Msg('stop', obj)` command registered on the RE by default (and a `bps.stop` plan stub). The semantics are the device should stop what ever it is doing, exactly how that gets translated to the underlying device is something the device class can control (some things push the `.STOP` field, some things set the setpoint to the current readback, etc).   This method is very local to the device, the device has no knowledge of what plan is currently being run so it can not meaningfully "move back to the beginning".  

`obj.stop()` is device specific method that semantics of a traffic cop holding up a hand in traffic: freeze your motion.

`RE.stop` is about the flow control of messages between the plan and the `RunEngine` (which is technically implemented by the RE throwing a `RequestStop` exception into the plan).  The semantics are wrapping up an experiment early because you have all the data you need.  Still do your clean up / close out process and mark the data as good.  It may intentionally call `obj.stop` as part of the plan and `RE()` will call `obj.stop()` on everything on the way out.

`RE.abort` works at the plan/message level and has the semantics of going home early because you discovered your only sample fell off the sample holder before you left home.  You still need to do the cleanup / close out process, but you mark the data as bad.  It may intentionally call `obj.stop` as part of the plan and `RE()` will call `obj.stop()` on everything on the way out.

`RE.halt` works at the plan level and implements the semantics of a  "stop-work-order".  The RE calls `obj.stop` on everything in the process of "stoping work"  but you do not do touch anything / clean up anything etc until you are given clearance to do so again.

**Comment by stan-dot** [comment]
there is lot of good explanations in this thread that could potentially go into some documentation
